src/raftstore/store/peer.rs

```diff
@@ -1058,15 +1157,108 @@ impl Peer {
                      quorum_after_change))
     }
 
+    fn transfer_leader(&mut self, peer: &metapb::Peer) {
+        info!("{} transfer leader to {:?}", self.tag, peer);
+
+        self.raft_group.transfer_leader(peer.get_id());
+    }
+
+    fn is_tranfer_leader_allowed(&self, peer: &metapb::Peer) -> bool {
+        let peer_id = peer.get_id();
+        let status = self.raft_group.status();
+
+        if !status.progress.contains_key(&peer_id) {
+            return false;
+        }
+
+        for progress in status.progress.values() {
+            if progress.state == ProgressState::Snapshot {
+                return false;
+            }
+        }
+
+        let last_index = self.get_store().last_index();
+        last_index <= status.progress[&peer_id].matched + TRANSFER_LEADER_ALLOW_LOG_LAG
+    }
+
+    fn read_local(&mut self, req: RaftCmdRequest, cb: Callback, metrics: &mut RaftProposeMetrics) {
+        metrics.local_read += 1;
+        self.handle_read(req, cb);
+    }
+
+    fn read_index(&mut self,
+                  uuid: Uuid,
+                  req: RaftCmdRequest,
+                  cb: Callback,
+                  metrics: &mut RaftProposeMetrics)
+                  -> bool {
+        metrics.read_index += 1;
+
+        let renew_lease_time = clocktime::raw_now();
+        if let Some(read) = self.pending_reads.reads.back_mut() {
+            if read.renew_lease_time + self.cfg.raft_store_max_leader_lease > renew_lease_time {
+                read.cmds.push((req, cb));
+                return false;
+            }
+        }
+
+        // Should we call pre_propose here?
+        let last_pending_read_count = self.raft_group.raft.pending_read_count();
+        let last_ready_read_count = self.raft_group.raft.ready_read_count();
+
+        self.raft_group.read_index(uuid.as_bytes().to_vec());
+
+        let pending_read_count = self.raft_group.raft.pending_read_count();
+        let ready_read_count = self.raft_group.raft.ready_read_count();
+
+        if pending_read_count == last_pending_read_count &&
+           ready_read_count == last_ready_read_count {
+            // The message gets dropped silently, can't be handled anymore.
+            notify_stale_req(&self.tag, self.term(), uuid, cb);
+            return false;
+        }
+
+        self.pending_reads.reads.push_back(ReadIndexRequest {
+            uuid: uuid,
+            cmds: vec![(req, cb)],
+            renew_lease_time: renew_lease_time,
+        });
+
+        match self.leader_lease_expired_time {
+            Some(Either::Right(_)) => {}
+            _ => return true,
+        };
+
+        // TimeoutNow has been sent out, so we need to propose explicitly to
+        // update leader lease.
+
+        let uuid = Uuid::new_v4();
+        if self.proposals.contains(&uuid) {
```

Comment:
> In contrast, with version 1 and 2 UUIDs using randomly-generated node ids, hash-based version 3 and 5 UUIDs, and random version 4 UUIDs, collisions can occur even without implementation problems, albeit with a probability so small that it can normally be ignored.

https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions